### DISCUSSION TITLE
Table-driven testing vs Data-driven testing

### DISCUSSION THREAD

**Discussion by Kantis** [kotest/kotest#4052] [6.0 changes]
As mentioned in [this PR comment](https://github.com/kotest/kotest/pull/4007#discussion_r1622711121), perhaps we should also have a look at "[table-driven tests](https://github.com/kotest/kotest/tree/master/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/data)" vs. "[data-driven tests](https://github.com/kotest/kotest/tree/master/kotest-framework/kotest-framework-datatest/src/commonMain/kotlin/io/kotest/datatest)". They're very similar and I think that makes it hard for users to know what to use.

I also think there's problems with forAll being used for both [property testing](https://github.com/kotest/kotest/blob/f4cd449df1cb5867941ff4e259c7f9333e5ba048/kotest-property/src/commonMain/kotlin/io/kotest/property/propertyTest1.kt#L89) and [table-driven tests](https://github.com/kotest/kotest/blob/f4cd449df1cb5867941ff4e259c7f9333e5ba048/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/data/forAll1.kt#L6).

I think it's important that there's not many ways to achieve the same thing with only slight nuances. It makes it hard for users to understand the framework and harder to maintain as well.

**Comment by sksamuel** [comment]
table driven tests is the precursor to data driven tests and if we need to remove something, it should be that.

**Reply by Kantis** [reply]
I think the only thing w/ regards to table-driven testing that we'd be missing is the table-from-file stuff that @jmfayard added in https://github.com/kotest/kotest/pull/3197

I'm leaning in favor of deprecating table-driven testing, and merging `kotest-framework-datatest` into some appropriate module.

**Reply by sksamuel** [reply]
Definitely we should merge datatest module into the engine
